wire-ios-sync-engine/Tests/Source/Calling/MLSConferenceStaleParticipantsRemoverTests.swift

```diff
@@ -0,0 +1,270 @@
+        // GIVEN
+
+        // create call participants
+        let participants = [
+            createMLSParticipant(state: .connecting),
+            createMLSParticipant(state: .connecting),
+            createMLSParticipant(state: connectedState),
+            createMLSParticipant(state: connectedState)
+        ]
+
+        // mock subconversation members
+        mlsService.mockSubconversationMembers = { _ in
+            return participants.map(\.mlsClientID)
+        }
+
+        // set expectations
+        let expectations = expectations(from: participants)
+        mlsService.mockRemoveMembersFromConversation = { clientIDs, _ in
+            guard let id = clientIDs.first else { return }
+            expectations[id]?.fulfill()
+        }
+
+        // WHEN
+        _ = sut.receive(
+            MLSConferenceParticipantsInfo(
+                participants: participants.map(\.callParticipant),
+                subconversationID: groupID
+            )
+        )
+
+        // THEN
+        wait(for: Array(expectations.values), timeout: 0.5)
+    }
+
+    func test_ItDoesntRemoveParticipantsThatReconnectedBeforeTimeout() {
+        // GIVEN
+
+        // create call participants
+        let participants = [
+            createMLSParticipant(state: .connecting),
+            createMLSParticipant(state: .connecting),
+            createMLSParticipant(state: .connecting)
+        ]
+
+        // mock subconversation members
+        mlsService.mockSubconversationMembers = { _ in
+            return participants.map(\.mlsClientID)
+        }
+
+        // WHEN
+        _ = sut.receive(
+            MLSConferenceParticipantsInfo(
+                participants: participants.map(\.callParticipant),
+                subconversationID: groupID
+            )
+        )
+
+        participants[0].updateState(connectedState)
+        participants[1].updateState(connectedState)
+
+        let expectations = expectations(from: participants)
+        mlsService.mockRemoveMembersFromConversation = { clientIDs, _ in
+            guard let id = clientIDs.first else { return }
+            expectations[id]?.fulfill()
+        }
+
+        _ = sut.receive(
+            MLSConferenceParticipantsInfo(
+                participants: participants.map(\.callParticipant),
+                subconversationID: groupID
+            )
+        )
+
+        // THEN
+        wait(for: Array(expectations.values), timeout: 0.5)
+    }
+
+    func test_ItDoesntRemoveStaleParticipantIfTheyAreNotInConversation() {
+        // GIVEN
+
+        // create call participants
+        let participants = [
+            createMLSParticipant(state: .connecting),
+            createMLSParticipant(state: .connecting)
+        ]
+
+        // mock subconversation members
+        mlsService.mockSubconversationMembers = { _ in
+            return []
+        }
+
+        // set expectations
+        let expectation = XCTestExpectation()
+        expectation.isInverted = true
```

Comment:
Interesting, I didn't know about this. So inverting it means we expect it not to be fulfilled within a certain timeout?

Reply:
Yep!